### ISSUE TITLE
referencedGrokFields does not detect all referenced fields

### ISSUE THREAD

**Issue by boriseveoh** [fstab/grok_exporter#10]
When a grok field is referenced in the following way:
```
{{if eq .field "value"}}text{{end}}
```
The field is not detected by `referencedGrokFields` (in metrics.go).

Example config:
```
global:
    config_version: 2
input:
    type: file
    path: ./input
    readall: true
grok:
    patterns_dir: ./patterns
metrics:
    - type: gauge
      name: example
      help: not empty
      match: '%{NOTSPACE:val1} %{NOTSPACE:val2}'
      value: '{{.val1}}'
      labels:
          my_label: '{{if eq .val2 "test"}}yes{{else}}no{{end}}'
```
With input file:
```
1 test
2 nomatch
```
This results in the following error:
```
WARNING: Skipping log line: unexpected error while evaluating my_label template: template: my_label:1:5: executing "my_label" at <eq .val2 "test">: error calling eq: invalid type for comparison
```

If we change the definition of `my_label` to:
```
my_label: '{{.val2}} {{if eq .val2 "test"}}yes{{else}}no{{end}}'
```
The metric works as expected:
```
# HELP example not empty
# TYPE example gauge
example{my_label="nomatch no"} 2
example{my_label="test yes"} 1
```

**Comment by fstab** [comment]
Thanks! I pushed a fix (see `templates.go` and `templates_test.go`), but need to review it before I build a new release.

**Comment by boriseveoh** [comment]
Thanks for your quick response and fix. The following test case using `else if` (which is already handled correctly right now) is relevant for my own use case:

```
		{
			// Issue #10
			template:           "{{if eq .field1 .field2}}{{.field3}}{{else if eq .field4 .field5}}{{.field6}}{{else}}{{.field7}}{{end}}",
			expectedGrokFields: []string{"field1", "field2", "field3", "field4", "field5", "field6", "field7"},
		},
```

It's still possible to come up with templates that won't work (eg. `{{define "T1"}} {{.field}} {{end}} {{template "T1" .}}` - the `define` block doesn't even show up in the parsed `Tree`), but I suppose templates this complex are simply "not supported".

**Comment by fstab** [comment]
Hi, thanks for the examples. I pushed some code and test for the `{{define "T1"}} {{.field}} {{end}}`. It seems to work, but I only tried one single example, so I'm not sure. Will look into it in more detail soon. If you want to try it, there's a new `TestExecute()` now in `template_test.go`. If you have any other examples, please go ahead :)

**Comment by boriseveoh** [comment]
I did come up with two more examples. Too bad there's not an easier way to iterate over all the nodes in the parsed template.
```
{{with $x := .field}} This is $x: {{$x}} {{end}}
```

```
{{define "T1"}} {{.}} {{end}} {{template "T1" .field}}
```

**Comment by fstab** [comment]
I rewrote the template part, it now tests the full list of actions included in Go's `text/template` documentation https://golang.org/pkg/text/template/. It works for your examples above and should work for all Go templates. Thanks for you examples, and please report back if there are still problems.

**Comment by fstab** [comment]
Released v0.2.1 with better Go template support. Closing this issue. If there are still problems, please open a new issue. Thanks for you feedback!
